extension/api.js

```diff
@@ -78,14 +105,174 @@ const lostEngines = new Map ([
+    return;
+  }
+
+  // Make sure the setting is not user controlled.
+  let control = await ExtensionSettingsStore.getLevelOfControl(
+    ExtensionSettingsStore.SETTING_USER_SET,
+    DEFAULT_SEARCH_STORE_TYPE,
+    DEFAULT_SEARCH_SETTING_NAME
+  );
+  if (control == "controlled_by_this_extension") {
+    console.log(`reset-default-search: search control has been user selected, not resetting, finished.`)
+    finish();
+    return;
+  }
+
+  // Get the latest installed addon that was installed prior to it's failure date.  Any non-app
+  // disabled state (ie. it is not disabled due to blocklisting) removes the addon from elibility.
+  let addons = (await AddonManager.getAddonsByIDs(
+      Array.from(lostEngines.keys())
+    )).filter(
+      a => a &&
+      !a.userDisabled &&
+      !a.softDisabled &&
+      !a.embedderDisabled &&
+      lostEngines.has(a.id) &&
+      a.installDate < lostEngines.get(a.id)
+    );
+  if (!addons.length) {
+    console.log("reset-default-search: No addons in our list are installed.");
+    finish();
+    return;
+  }
+
+  // Filter out any that are blocklisted.
+  let enabledAddons = addons.filter(a => !a.appDisabled);
+  console.log(`enabledAddons ${enabledAddons.length}`);
+
+  // We will only ask for the latest installed engine.  We will
+  // loop through the list until one of them makes it to the prompt.
+  enabledAddons.sort((a, b) => b.installDate - a.installDate);
+  for (let addon of enabledAddons) {
+    console.log(`reset-default-search: reset search engine to ${addon.id}`);
+
+    let policy = WebExtensionPolicy.getByID(addon.id);
+    if (!policy?.extension) {
+      tryagain = true;
+      console.log("reset-default-search: extension is not running, cannot set as default, try again later");
+      continue;
+    }
+    let { extension } = policy;
+    let { manifest } = extension;
+
+    let searchProvider = manifest?.chrome_settings_overrides?.search_provider;
+    if (!searchProvider?.is_default) {
+      // If the extension isn't asking to be default at this point, bail out.
+      console.log("reset-default-search: is_default is not requested by the addon, try again later");
+      continue;
+    }
+
+    let engineName = searchProvider.name.trim();
+    if (!Services.search.getEngineByName(engineName)) {
+      // attempt to load the engine, but if it fails then we'll try again later.
+      console.log(`reset-default-search: engine is not configured in search, try adding ${policy.extension.id}`);
+      // "default" here is about the locale, not default setting.  We dig into search service here due
+      // to a bug (fixed in 89) that prevents using addEnginesFromExtension for updates.
+      await Services.search.wrappedJSObject._installExtensionEngine(extension, ["default"]);
+      if (!Services.search.getEngineByName(engineName)) {
+        tryagain = true;
+        console.log(`reset-default-search: could not configure ${policy.extension.id} in search, try again later`);
```

Comment:
Use `addon.id` here (or even `extension.id`), but not `policy.extension.id`. `policy.extension` can be null when the extension has been unloaded. Earlier, you're checking for `!policy?.extension`, but there is an `await` up there during which the extension can be unloaded.

You've also used `policy.extension.id` up there, which is currently safe due to the `!policy?.extension` check, but let's also use `addon.id` or `extension.id` there fore consistency.